rir/src/compiler/pir/instruction.h

```diff
@@ -498,18 +499,31 @@ class FLI(LdConst, 0, Effect::None, EnvAccess::None) {
     void printArgs(std::ostream& out, bool tty) override;
 };
 
-class FLIE(LdFun, 1, Effect::Any, EnvAccess::Write) {
+class FLIE(LdFun, 2, Effect::Any, EnvAccess::Write) {
   public:
     SEXP varName;
 
     LdFun(const char* name, Value* env)
-        : FixedLenInstructionWithEnvSlot(RType::closure, env),
+        : FixedLenInstructionWithEnvSlot(RType::closure, {{PirType::any()}},
+                                         {{Tombstone::closure()}}, env),
           varName(Rf_install(name)) {}
     LdFun(SEXP name, Value* env)
-        : FixedLenInstructionWithEnvSlot(RType::closure, env), varName(name) {
+        : FixedLenInstructionWithEnvSlot(RType::closure, {{PirType::any()}},
+                                         {{Tombstone::closure()}}, env),
+          varName(name) {
         assert(TYPEOF(name) == SYMSXP);
     }
 
+    void clearGuessedBinding() { arg<0>().val() = Tombstone::closure(); }
+
+    void guessedBinding(Value* val) { arg<0>().val() = val; }
```

Comment:
FWIW I'm fine with `guessedBinding`, but more comments could help.

Reply:
> the binding may be a temporal value we are not a 100% sure, but after the whole analysis passes if it has a binding we are sure

unfortunately not. Let me give you an annotated example:

```
test <- function(f) f()
test(function()1)
```
so in a kind of PIR syntax this is:
```
test:
  %t1 = ldarg 1
  %t2 = mkenv "f"=%t1
  %t3 = ldfun "f", %t2
        call %t3

main:
  %m1 = mkarg(p1)
  %m2 = ldfun "test"
        call %m2(%m1)
 prom1:
  return mkClosure(...)
```

ok, so first step is of course to inline test into main, so we get:

```
main:
  %1 = mkarg(p1)
  %2 = mkenv "f"=%1
  %3 = ldfun "f", %2
        call %3
 prom1:
  return mkClosure(...)
```

so now we would like to inline the promise `p1`, but how? Our promise inliner pass matches up `mkArg <--> force` pairs and inlines the promise at the location of the force instruction (because that's where it get's actually executed). But as you can see here, there just isn't any force instruction, because of the weird semantics of lookup, the "f" is forced as a sideeffect of `%t3 = ldfun "f", %2`.

How to remedy?

*Step1:* get the desired force instruction

The scope resolution pass creates the following:

```
main:
  %1   = mkarg(p1)
  %2   = mkenv "f"=%1
  %3.1 = force %1
  %3   = ldfun "f", %2, guess = %3.1
          call %3
 prom1:
  return mkClosure(...)
```

(Note, here I simplified to much. Actually the return type of 'prom1' is trivially 'closure', so we would not need to guess at all. But imagine it is just `return (ldvar x)`, we might have no clue at all, what the guess is.)

*Step2:* promise inliner

Voila, we have a force (`%t3.1`) and a mkarg (`%1`) that match up. We can inline the promise code:

```
main:
  %1   = mkarg(p1)
  %2   = mkenv "f"=%1
  %3.1 = mkClosure(...)
  %3   = ldfun "f", %2, guess = %3.1
          call %3
```

*Step3:* promote the guess

Well, now it's easy. `%3.1` has type closure, and so we know that the guess was right. We will replace `%3` by `%3.1` and remove `%3`.

But, there are two other possible outcomes:

1. The guess is wrong:

after promise inlining we might end up with:

 ```
main:
  %1   = mkarg(p1)
  %2   = mkenv "f"=%1
  %3.1 = ldConst 1
  %3   = ldfun "f", %2, guess = %3.1
          call %3
```

so we know the guess was wrong since the innermost binding is just the number 1. This happens if I would call test like this: `f <- function()1; test(1)`.

What we could do in this case (we don't yet) is to replace the guess with the "next best guess" (ie. the next binding).

2. We can't show the guess to be true nor false

after promise inlining we might end up with:

 ```
main:
  %1   = mkarg(p1)
  %2   = mkenv "f"=%1
  %3.1 = call("mystery")
  %3   = ldfun "f", %2, guess = %3.1
          call %3
```

The first binding is something we have no clue what it is. We can neither exclude it, since mystery might return a closure, nor show it will be skipped, since mystery might not return a closure.

In this case the backend of the compiler just forgets the whole guess business. When we lower to rir, the guess is dropped on the floor.

Reply:
Thanks a lot for the explanation!

Although I stated it wrong, your explanation reinforces somehow what I wanted to mean.
The name guess resorts me to a dynamic setting, in which I would execute code, and if the guess was wrong I must act accordingly.
However, this behavior happens within the compiler statically. The argument of the `ldFun` is just a static binding of the name at the point where this instruction would be executed. IMO that is much more clear for somebody reading the code. Then in the lowering phase, you work with the semantics of the ldFun: if we can not ensure the binding is a closure we can not change the `ldFun` for a simple call. 

Summarizing: IMO the naming is misleading. Since it encapsulates a lot of assumptions about why it is actually there for, I thought it was important to discuss. But you and @mhyee think it is the correct one so leave the current naming.